### PR TITLE
Enhance merged anomaly calculation

### PR DESCRIPTION
**Pull request by cyenjung**
1. Add force backfill flag, which indicate this backfill session does not want to resume from any previous backfill sessions.

2. Create a method that uses anomaly function specific method to update weight, whose absolute value is severity, of merged anomalies. If the update is failed, then the average of its raw anomalies is used.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/merge/AnomalyMergeExecutor.java [hunk 1]
```diff
@@ -184,128 +179,55 @@ private void updateMergedScoreAndPersist(MergedAnomalyResultDTO mergedResult) {
     }
   }
 
-  // TODO: Update severity according to history data
-  private void updateMergedSeverity(MergedAnomalyResultDTO anomalyMergedResult) throws Exception {
+  private void updateMergedAnomalyWeight(MergedAnomalyResultDTO anomalyMergedResult) throws Exception {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/merge/AnomalyMergeExecutor.java]: Huge improvement here in terms of readability.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/merge/AnomalyMergeExecutor.java [hunk 2]
```diff
@@ -184,128 +179,55 @@ private void updateMergedScoreAndPersist(MergedAnomalyResultDTO mergedResult) {
...
 
-    timeSeriesRequest.setMetricExpressions(metricExpressions);
-    TimeGranularity timeBucket = new TimeGranularity(anomalyFunctionSpec.getBucketSize(),
+    TimeGranularity timeGranularity = new TimeGranularity(anomalyFunctionSpec.getBucketSize(),
         anomalyFunctionSpec.getBucketUnit());
-    timeSeriesRequest.setAggregationTimeGranularity(timeBucket);
 
-    timeSeriesRequest.setEndDateInclusive(false);
+    MetricTimeSeries metricTimeSeries =
+        TimeSeriesUtil.getTimeSeriesResponseForPresentation(anomalyFunctionSpec, startEndTimeRanges,
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/merge/AnomalyMergeExecutor.java]: if we are using this for both presentation(UI) and background job, we should rename getTimeSeriesResponseForPresentation method

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DashboardResource.java [hunk 3]
```diff
@@ -558,4 +565,29 @@ public String getData(@QueryParam("type") String type) throws Exception {
     }
     return response;
   }
+
+  @POST
+  @Path("/severity")
+  public Response computeSeverity(@NotNull @QueryParam("collection") String collectionName,
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DashboardResource.java]: suggest moving this to another Resource, may be DetectionJobResource or even Admin is also fine.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DashboardResource.java [hunk 4]
```diff
@@ -558,4 +565,29 @@ public String getData(@QueryParam("type") String type) throws Exception {
...
+      throws Exception {
+    DateTime startTime = null;
+    DateTime endTime = null;
+    if (StringUtils.isNotBlank(startTimeIso)) {
+      startTime = ISODateTimeFormat.dateTimeParser().parseDateTime(startTimeIso);
+    }
+    if (StringUtils.isNotBlank(endTimeIso)) {
+      endTime = ISODateTimeFormat.dateTimeParser().parseDateTime(endTimeIso);
+    }
+    PinotThirdEyeClient thirdEyeClient = PinotThirdEyeClient.getDefaultTestClient();
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DashboardResource.java]: this was for testing. You should probably get the thirdeyePinotClient from else where or pass it when resource is getting created

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DashboardResource.java [hunk 5]
```diff
@@ -558,4 +565,29 @@ public String getData(@QueryParam("type") String type) throws Exception {
...
+    SeverityComputationUtil util = new SeverityComputationUtil(thirdEyeClient, collectionName, metricName);
+    long currentWindowStart = startTime.getMillis();
+    long currentWindowEnd = endTime.getMillis();
+    String compareMode = "WO3WMean";
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DashboardResource.java]: add a TODO to support more compare Mode. Also with the new change to sign test algo, we might want wo4wMean right?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/function/AnomalyFunction.java [hunk 6]
```diff
@@ -37,6 +49,24 @@
...
+   *          the merged anomaly to be updated.
+   * @param timeSeries
+   *          The metric time series data.
+   * @param knownAnomalies
+   *          Any known anomalies in the time range.
+   * @return
+   *         A list of anomalies that were not previously known.
+   * @return the severity according to the current and baseline of the given timeSeries
+   */
+  void updateMergedAnomalyInfo(MergedAnomalyResultDTO anomalyToUpdated, MetricTimeSeries timeSeries,
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/function/AnomalyFunction.java]: will timeseries have data from start to end?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/util/SeverityComputationUtil.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      intervals.add(ImmutablePair.of(new DateTime(currentWindowStart).minusDays(7), new DateTime(currentWindowEnd).minusDays(7)));
+      intervals.add(ImmutablePair.of(new DateTime(currentWindowStart).minusDays(14), new DateTime(currentWindowEnd).minusDays(14)));
+      intervals.add(ImmutablePair.of(new DateTime(currentWindowStart).minusDays(21), new DateTime(currentWindowEnd).minusDays(21)));
+      return intervals;
+    }
+    return Collections.emptyList();
+
+  }
+
+  public static void main(String[] args) throws Exception {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/util/SeverityComputationUtil.java]: remove this or take the params as arguments

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/function/AnomalyFunction.java (hunk 6 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/function/AnomalyFunction.java]: It has time keys, but not start and end of the values of current or baselines. The assumption is that the time series have all the time keys that are needed for calculating the weight of the merged anomaly.

Actually, the start and end time can be retrieved from the merged anomaly, but I want to make the signature of this function similar to that of analyze() method.

**Pull request merged**